crates/matrix-sdk/src/account.rs

```diff
@@ -812,6 +813,18 @@ impl Account {
     pub async fn mark_as_dm(&self, room_id: &RoomId, user_ids: &[OwnedUserId]) -> Result<()> {
         use ruma::events::direct::DirectEventContent;
 
+        // This function does a read/update/store of an account data event stored on the
+        // homeserver. We first fetch the existing account data event, the event
+        // contains a map which gets updated by this method, finally we upload the
+        // modified event.
+        //
+        // To prevent multiple calls to this method trying to update the map of DMs same
+        // time, and thus trampling on each other we introduce a lock which acts
+        // as a semaphore.
+        static LOCK: Mutex<()> = Mutex::const_new(());
```

Comment:
uh, a static lock will make it impossible for two different client instances to run this functions at the same time, right? could we use a lock that's per-client instance instead?

Reply:
Oh duh, that's a good point. :facepalm:

Reply:
This should help: https://github.com/matrix-org/matrix-rust-sdk/pull/2637